packages/realm/src/schema/normalize.ts

```diff
@@ -43,163 +51,407 @@ export const PRIMITIVE_TYPES = new Set<PropertyTypeName>([
   if (typeof arg === "function") {
     assert.extends(arg, RealmObject);
     assert.object(arg.schema, "schema static");
     const schema = normalizeObjectSchema(arg.schema as ObjectSchema);
-    schema.constructor = arg;
+    schema.ctor = arg;
     return schema;
-  } else {
-    // TODO: Determine if we still want to support this
-    if (Array.isArray(arg.properties)) {
-      if (flags.ALLOW_VALUES_ARRAYS) {
-        return normalizeObjectSchema({
-          ...arg,
-          properties: Object.fromEntries(arg.properties.map(({ name, ...rest }) => [name, rest])),
-        });
-      } else {
-        throw new Error("Array of properties are no longer supported");
-      }
+  }
+
+  // ---- THIS IF BLOCK HAS NOT YET BEEN REWRITTEN ----
+  // TODO: Determine if we still want to support this (should show warning to users of future deprecation)
+  if (Array.isArray(arg.properties)) {
+    if (flags.ALLOW_VALUES_ARRAYS) {
+      return normalizeObjectSchema({
+        ...arg,
+        // Build the PropertiesTypes object
+        properties: Object.fromEntries(arg.properties.map(({ name, ...rest }) => [name, rest])),
+      });
     }
-    return {
-      constructor: undefined,
-      name: arg.name,
-      primaryKey: arg.primaryKey,
-      asymmetric: arg.asymmetric || false,
-      embedded: arg.embedded || false,
-      properties: Object.fromEntries(
-        Object.entries(arg.properties).map(([name, property]) => {
-          const canonicalPropertySchema = normalizePropertySchema(name, property);
-          // A primary key is always indexed
-          if (name === arg.primaryKey) {
-            canonicalPropertySchema.indexed = true;
-          }
-          validateCanonicalPropertySchema(arg.name, canonicalPropertySchema);
-          return [name, canonicalPropertySchema];
-        }),
-      ),
-    };
+    throw new Error("Array of properties are no longer supported. Use an object instead.");
   }
+  // --------------------------------------------------
+
+  return {
+    name: arg.name,
+    primaryKey: arg.primaryKey,
+    asymmetric: !!arg.asymmetric,
+    embedded: !!arg.embedded,
+    properties: normalizePropertySchemas(arg.name, arg.properties, arg.primaryKey),
+  };
+}
+
+/**
+ * Transform user-provided property schemas into their canonical forms.
+ */
+function normalizePropertySchemas(
+  objectName: string,
+  schemas: PropertiesTypes,
+  primaryKey?: string,
+): Record<string, CanonicalObjectSchemaProperty> {
+  const normalizedSchemas: Record<string, CanonicalObjectSchemaProperty> = {};
+  for (const propertyName in schemas) {
+    normalizedSchemas[propertyName] = normalizePropertySchema(
+      { objectName, propertyName },
+      schemas[propertyName],
+      primaryKey === propertyName,
+    );
+  }
+
+  return normalizedSchemas;
 }
 
-// TODO: Extend to handle all short hands
+/**
+ * Transform a user-provided property schema into its canonical form.
+ */
 export function normalizePropertySchema(
```

Comment:
We currently treat everything as case-sensitive. Thus, `"Int"` will be interpreted as a user-defined type. I don't have a strong opinion in this regard, but don't see an issue keeping it case-sensitive for now.

Reply:
I commented elsewhere on this, but in my opinion things should be case-sensitive.  I think there shouldn't be too many ways of defining the same thing, unless that somehow provides value, which I can't really think of.  What we should do, is help our users do the right thing.  

We have talked about enumerating all the primitive types, but this would make the shorthand less valuable.  

We also have the babel-plugin, which takes a typescript class and infers the entire schema from that.  This leverages typescript syntax instead of a shorthand notation, which is also quite valuable.

In any case, we should have a broader discussion about this.  I think there are many improvements we can have to make defining data models more idiomatic and user friendly.

Reply:
Good points!